cmd/libsnap-confine-private/cgroup-support-test.c

```diff
@@ -0,0 +1,284 @@
+    g_file_set_contents(fixture->p, "0::/foo/bar/baz/snap.foo.app.1234-1234.scope", -1, &err);
+    g_assert_no_error(err);
+
+    /* our group is the only one for this snap */
+    const char *dirs[] = {
+        "/foo/bar/baz/snap.foo.app.1234-1234.scope",
+        "/foo/bar/bad",
+        "/system.slice/some/app/other",
+        "/user/slice/other/app",
+    };
+
+    for (size_t i = 0; i < sizeof dirs / sizeof dirs[0]; i++) {
+        g_autofree const char *np = g_build_filename(fixture->root, dirs[i], NULL);
+        int ret = g_mkdir_with_parents(np, 0755);
+        g_assert_cmpint(ret, ==, 0);
+    }
+
+    bool is_tracking = sc_cgroup_v2_is_tracking_snap("foo");
+    /* our own group is skipped */
+    g_assert_false(is_tracking);
+}
+
+static void test_sc_cgroupv2_is_tracking_no_dirs(cgroupv2_is_tracking_fixture *fixture, gconstpointer user_data) {
+    GError *err = NULL;
+    g_file_set_contents(fixture->p, "0::/foo/bar/baz/snap.foo.app.scope", -1, &err);
+    g_assert_no_error(err);
+
+    bool is_tracking = sc_cgroup_v2_is_tracking_snap("foo");
+    g_assert_false(is_tracking);
+}
+
+static void test_sc_cgroupv2_is_tracking_bad_self_group(cgroupv2_is_tracking_fixture *fixture,
+                                                        gconstpointer user_data) {
+    GError *err = NULL;
+    /* trigger a failure in own group handling */
+    g_file_set_contents(fixture->p, "", -1, &err);
+    g_assert_no_error(err);
+
+    if (g_test_subprocess()) {
+        sc_cgroup_v2_is_tracking_snap("foo");
+    }
+    g_test_trap_subprocess(NULL, 0, 0);
+    g_test_trap_assert_failed();
+    g_test_trap_assert_stderr("cannot obtain own group path\n");
+}
+
+static void test_sc_cgroupv2_is_tracking_dir_permissions(cgroupv2_is_tracking_fixture *fixture,
+                                                         gconstpointer user_data) {
+    if (geteuid() == 0) {
+        g_test_skip("the test will not work when running as root");
+        return;
+    }
+    GError *err = NULL;
+    g_file_set_contents(fixture->p, "0::/foo/bar/baz/snap.foo.app.1234-1234.scope", -1, &err);
+    g_assert_no_error(err);
+
+    /* there exist 2 groups with processes from a given snap */
+    const char *dirs[] = {
+        "/foo/bar/bad",
+        "/foo/bar/bad/badperm",
+    };
+    for (size_t i = 0; i < sizeof dirs / sizeof dirs[0]; i++) {
+        int mode = 0755;
+        if (g_str_has_suffix(dirs[i], "/badperm")) {
+            mode = 0000;
+        }
+        g_autofree const char *np = g_build_filename(fixture->root, dirs[i], NULL);
+        int ret = g_mkdir_with_parents(np, mode);
+        g_assert_cmpint(ret, ==, 0);
+    }
+
+    /* dies when hitting an error traversing the hierarchy */
+    if (g_test_subprocess()) {
+        sc_cgroup_v2_is_tracking_snap("foo");
+    }
+    g_test_trap_subprocess(NULL, 0, 0);
+    g_test_trap_assert_failed();
+    g_test_trap_assert_stderr("cannot open directory entry \"badperm\": Permission denied\n");
```

Comment:
(nipick) maybe also some information here that this is cgrop v2 code related? e.g. `cannot open cgroup v2 directory entry ... ` ? But of course this (and my suggestion above) are very subjective so I don't mind if they are changed or ignored